src/coreComponents/physicsSolvers/fluidFlow/kernels/immiscibleMultiphase/ImmiscibleMultiphaseKernels.hpp

```diff
@@ -0,0 +1,1420 @@
+            }
+
+            if( m_hasCapPressure )  // check sign convention
+            {
+              real64 const capPres = m_phaseCapPressure[er][esr][ei][0][ip];  // Pc = Pc1 || Pc2
+              dCapGrad_dTrans -= signPotDiff[ke] * capPres;                   // dDPc/dT = (-Pc1 + Pc2)
+              pot -= trans[ke] * capPres;                                     // Phi = T P1 - rho T g z1 - T Pc1 || -T P2 + rho T g z2 + T
+                                                                              // Pc2
+              capGrad[ip] -= trans[ke] * capPres;                             // DPc = T (-Pc1 + Pc2)
+            }
+
+            potScale = fmax( potScale, fabs( pot ) ); // maxPhi = Phi1 > Phi2 ? Phi1 : Phi2
+          }
+
+          for( integer ke = 0; ke < 2; ++ke )
+          {
+            dPresGrad_dP[ip][ke] += dTrans_dP[ke] * dPresGrad_dTrans;   // dDPv/dP = { T + dT/dP1 * (P1 - P2) , -T + dT/dP2 * (P1 - P2)}
+            dGravHead_dP[ip][ke] += dTrans_dP[ke] * dGravHead_dTrans;   // dDPg/dP = { drho/dP1 * T g (z1 - z2) + dT/dP1 * rho g (z1 - z2) ,
+                                                                        //             drho/dP2 * T g (z1 - z2) + dT/dP2 * rho g (z1 - z2) }
+            if( m_hasCapPressure )
+            {
+              real64 const dCapPres_dS = m_dPhaseCapPressure_dPhaseVolFrac[seri[ke]][sesri[ke]][sei[ke]][0][ip][ip]; // dPc/dS = dPc1/dS1 ||
+                                                                                                                     // dPc2/dS2
+              dCapGrad_dP[ip][ke] += dTrans_dP[ke] * dCapGrad_dTrans;                                                // dDPc/dP = { dT/dP1 *
+                                                                                                                     // (-Pc1 + Pc2) ,
+                                                                                                                     //             dT/dP2 *
+                                                                                                                     // (-Pc1 + Pc2) }
+              dCapGrad_dS[ip][ke] -= trans[ke] * dCapPres_dS;                                                        // dDPc/dS = { -T *
+                                                                                                                     // dPc1/dS1 , T *
+                                                                                                                     // dPc2/dS2 }
+            }
+          }
+
+          // *** upwinding ***
+
+          // compute potential gradient
+          real64 potGrad = presGrad[ip] - gravHead[ip]; // DPhi = T (P1 - P2) - T rho g (z1 - z2)
+          if( m_hasCapPressure )
+          {
+            potGrad += capGrad[ip]; // DPhi = T (P1 - P2) - T rho g (z1 - z2) + T (-Pc1 + Pc2)
+          }
+
+          // compute upwinding tolerance
```

Comment:
compositional is not doing that trick but ok to keep it here as long as it works fine